### DISCUSSION TITLE
DIRAC.rootPath issues

### DISCUSSION THREAD

**Discussion by sfayer** [DIRACGrid/DIRAC#5991] [Q&A] [answered]
Hi,

We've had a long standing problem with DIRAC.rootPath pointing to the wrong directory on our server installs... We're using a classic non-pro style layout (closer to a modern developer install: We do this so we have git history to track upgrades/hotpaches/tweaks/etc.):
```
$ ls /opt/dirac
bashrc data DIRAC diracos etc runit startup work
```

We'd ultimately like `DIRAC.rootPath=/opt/dirac`: We could move the directories (like etc) around but having them at the top level keeps them safe. For example if we had to move them into /opt/dirac/diracos, we consider that to be expendable: It's convenient to be able to just rm -Rf the whole thing and recreate it on major updates without having to pick out the bits like etc/dirac.cfg we'd need to keep.

The default value of rootPath this has jumped around a few times for our layout over the last year or two:
```
print(DIRAC.rootPath)
v7r0(py2): /opt/dirac
v7r1(py2): /opt/dirac
v7r2(py2): /opt/dirac/DIRAC_src/src/DIRAC (We did some symlink tricks here to get things to work)
v7r3(py2): /opt/dirac/DIRAC_src/src/DIRAC
v7r3(py3): /opt/dirac/diracos
```
Up until now we've just been hotpatching `__init__.py` to override rootPath with the value of $DIRAC, but now we're testing python3 we'd like to find a proper fix (it's a bit tedious to remember, we frequently forget to stash pop the patch on updates).

Is there some way we can get the rootPath detected correctly in our layout? At the very least perhaps we could consider adding a new environment variable called DIRACTOP [*] or something that's used instead of the detection logic if it's set? 

[*] I don't think there is any existing variable suitable for this, even $DIRAC seems to point at /opt/dirac/diracos on a py3 server instance. My initial suggestion was going to be DIRACROOT, but this is used elsewhere and I don't want to risk breaking anything.

Regards,
Simon

**Comment by chrisburr** [comment]
I think you can achive this by setting `DIRACSYSCONFIG` in the environment to point to an alternative `dirac.cfg` location that then contains keys like:

```
/LocalInstallation/RunitDir
/LocalInstallation/StartupDir
```

**Reply by sfayer** [reply]
Thanks, we've had a look at this, but we don't think it'll quite work. There are instances where rootPath is used directly with no override option:
https://github.com/DIRACGrid/DIRAC/blob/7859236ca8f81c83240fb37af2df60d100992925/src/DIRAC/ConfigurationSystem/private/ConfigurationData.py#L32

I'd be a bit dubious of having to set lots of variables correctly when the root cause is rootPath being detected wrong: That feels fragile in-case any new variables are added in the future.

Regards,
Simon

**Reply by chrisburr** [reply]
Personally I think I'd rather have an option in `dirac.cfg` to override this path, espeically with the view to eventually running DIRAC in short-lived containers rather than with `runsvdir`.

Maybe @chaen has thoughts given he's been playing with kubernetes?

**Reply by chaen** [reply]
however problematic they are, environment variables play a bit more nicely with kubernetes :man_shrugging: 
Moreover, with the `dirac.cfg`, you will always have the problem of needing it when it is not yet initialized

**Comment by chrisburr** [comment, marked as answer]
Support for overriding this by setting the `DIRAC_ROOT_PATH` environment variable was added in https://github.com/DIRACGrid/DIRAC/pull/6011.
